examples/pytorch/infograph/semisupervised.py

```diff
@@ -0,0 +1,222 @@
+
+    # Step 1: Prepare graph data   ===================================== #
+    args = argument()
+    label_keys = [args.target]
+    print(args)
+
+    dataset = QM9DatasetV2(label_keys)
+    dataset.to_dense()
+
+    graphs = dataset.graphs
+
+    # Train/Val/Test Splitting
+    N = len(graphs)
+    all_idx = np.arange(N)
+    np.random.shuffle(all_idx)
+
+    val_num = 10000
+    test_num = 10000
+
+    val_idx = all_idx[:val_num]
+    test_idx = all_idx[val_num : val_num + test_num]
+    train_idx = all_idx[val_num + test_num : val_num + test_num + args.train_num]
+
+    train_data = [dataset[i] for i in train_idx]
+    val_data = [dataset[i] for i in val_idx]
+    test_data = [dataset[i] for i in test_idx]
+
+    unsup_idx = all_idx[val_num + test_num:]
+    unsup_data = [dataset[i] for i in unsup_idx]
+
+    # generate supervised training dataloader and unsupervised training dataloader
+    train_loader = GraphDataLoader(train_data,
+                                   batch_size=args.batch_size,
+                                   collate_fn=collate,
+                                   drop_last=False,
+                                   shuffle=True)
+
+    unsup_loader = GraphDataLoader(unsup_data,
+                                   batch_size=args.batch_size,
+                                   collate_fn=collate,
+                                   drop_last=False,
+                                   shuffle=True)
+
+    # generate validation & testing dataloader
+
+    val_loader = GraphDataLoader(val_data,
+                                 batch_size=args.val_batch_size,
+                                 collate_fn=collate,
+                                 drop_last=False,
+                                 shuffle=True)
+
+    test_loader = GraphDataLoader(test_data,
+                                  batch_size=args.val_batch_size,
+                                  collate_fn=collate,
+                                  drop_last=False,
+                                  shuffle=True)
+
+    print('======== target = {} ========'.format(args.target))
+
+    mean = dataset.labels.mean().item()
+    std = dataset.labels.mean().item()
+
+
+    print('mean = {:4f}'.format(mean))
+    print('std = {:4f}'.format(std))
+
+    in_dim = dataset[0][0].ndata['attr'].shape[1]
+
+    # Step 2: Create model =================================================================== #
+    model = InfoGraphS(in_dim, args.hid_dim)
+    model = model.to(args.device)
+
+    # Step 3: Create training components ===================================================== #
+    optimizer = th.optim.Adam(model.parameters(), lr=args.lr, weight_decay=args.wd)
+    scheduler = th.optim.lr_scheduler.ReduceLROnPlateau(
+        optimizer, mode='min', factor=0.7, patience=5, min_lr=0.000001
+    )
+
+    # Step 4: training epochs =============================================================== #
+    sup_loss_all = 0
+    unsup_loss_all = 0
+    consis_loss_all = 0
```

Comment:
The three lines are redundant as you redeclare them at the beginning of each epoch.